src/jmh/java/com/amazon/corretto/crypto/provider/AesGcmBase.java

```diff
@@ -0,0 +1,39 @@
+package com.amazon.corretto.crypto.provider;
+
+import java.security.Security;
+import java.security.Key;
+
+import javax.crypto.Cipher;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+import org.bouncycastle.jce.provider.BouncyCastleProvider;
+
+
+public class AesGcmBase {
+    protected static final int PLAINTEXT_SIZE = 1024 * 1024;
+
+    protected Key key;
+    protected GCMParameterSpec params1;
+    protected GCMParameterSpec params2;
+    protected Cipher encryptor;
+    protected Cipher decryptor;
+    protected byte[] plaintext;
+    protected byte[] ciphertext;
+
+    protected void setup(int keyBits, String provider) throws Exception {
+        BenchmarkUtils.setupProvider(provider);
+        key = new SecretKeySpec(BenchmarkUtils.getRandBytes(keyBits/8), "AES");
+        params1 = new GCMParameterSpec(128, BenchmarkUtils.getRandBytes(12));
+        params2 = new GCMParameterSpec(128, BenchmarkUtils.getRandBytes(12));
+        final String algorithm = "AES/GCM/NoPadding";
+        encryptor = Cipher.getInstance(algorithm, provider);
+        decryptor = Cipher.getInstance(algorithm, provider);
+        encryptor.init(Cipher.ENCRYPT_MODE, key, params1);
```

Comment:
What's the purpose of doing one round of encryption/decryption in the setup?

Reply:
we don't actually do any decryption in the setup. we encrypt the plaintext to pre-generate the ciphertext on L35. this way, the decrypt benchmarks have a ciphertext ready to go.

i guess i don't technically need the decryptor init call on L33. it's been a few months since i originally wrote this, i'm guessing that i included it for aesthetic reasons -- symmetry, completeness, etc.

IIRC i needed 2 different params objects because JCA won't let you re-init the same Cipher cipher instance with the same params object back-to-back.